### ISSUE TITLE
Problems with grading

### ISSUE THREAD

**Issue by cigas** [RunestoneInteractive/RunestoneComponents#437]
There are problems at runestone.academy with grading. I have an assignment that has a reading component and some exercises that I wrote myself and added using search. I was able to autograde the reading assignment. When I look at the student summary (below) it omits the exercises from the display, but includes the 10 points in the total. This is very confusing.

![grading](https://user-images.githubusercontent.com/11619455/29416725-a523c0b0-832c-11e7-911f-7b1b91cf30f5.PNG)

Furthermore, when I go to grade the exercises, some of them do not display. I see the students' code for  CS152_Prints_2 and CS152_Prints_3, but CS152_Prints_1 shows "Preview not available for null", even though this is just a one-line print statement that most students did in class.

![grading2](https://user-images.githubusercontent.com/11619455/29417218-0ea946c6-832e-11e7-98f0-d4fd50594ddd.PNG)

**Comment by bnmnetp** [comment]
OK, I'll look into why its not displaying the Problems...  If my memory is correct it was before.  Can you tell me the name of the course?  I assume this is not your test course anymore.

As to the one problem not showing -- Preview not available -- I have a guess, please tell me if I'm right.  Because that problem, which your wrote yourself, was so simple you decided you didn't need to preview it.  Then When I prompted you as to whether you previewed or not you lied to the computer and said that you had.  Previewing is necessary to generate the HTML that gets saved in the database for previewing and grading. -- maybe a design choice I need to reconsider.

**Comment by cigas** [comment]
The course is CS152HOFA2017.

It is possible that I did not preview it, but I know for a fact that I did preview the first problem I wrote, at least the first time. The confusion comes from editing. It is quite possible that I edited the problem and then did not preview it. 

Rather than call it preview, maybe you need to call it Generate View or Generate Problem or Generate Image or something similar. Or the explanation needs to say why this is necessary. 

Really though, I'd think you would want to do this automatically whenever an exercise is saved. I can't imagine the load being anywhere near what it is from all the rebuilds I do in response to prompts that there are bug fixes (which is a GREAT thing to tell people - I'm not complaining about that).

I'm also pretty jaded by the preview thing. I deal with a system at work that has an option to generate a preview. On my interface,  the preview image can be out of sync with what is actually stored, so I always view what is actually stored, since looking at the preview generated by someone else can be a huge waste of time.

**Comment by bnmnetp** [comment]
I could definitely do it on the server side, but I really want to make sure that instructors are not introducing error filled exercises into their assignments.  So, I've changed the language on the buttons, and added a simple check to make sure that the html is there before I allow someone to finally save and move on.

**Comment by bnmnetp** [comment]
For the questions you wrote yourself, it looks like the "autograde" and "which to grade" fields are null.  What do they look like in the assignment interface?  That **may** be why they are missing from the detail view... Still working on it.

**Comment by cigas** [comment]
When I built the Assignment, the Autograde and Which to Grade links gave me a menu with no items in it, so I couldn't set them to anything.  I see now the menus are populated. I can change them if you want, but will wait to hear back.

When I add a new exercise to an assignment, those two fields are initially Empty (see image). You may want to consider an an actual value as a default.

![newexercise](https://user-images.githubusercontent.com/11619455/29427841-b65bc804-8350-11e7-9bf9-ce8a9f1d0d65.PNG)

**Comment by bnmnetp** [comment]
I agree.

**Comment by bnmnetp** [comment]
Just be be sure I understand.  Did you create these new questions by writing them from scratch -- i.e. clicking on the Write button.  Or did you search for a question, then edit it to do what you wanted it to do?  I how have decent defaults showing up for both.

**Comment by cigas** [comment]
I initially did Write, and got a template with all the options, which I didn't know what to do with. I edited that many times, including copying in code from some other exercises in the book. I'm pretty sure at some point I figured out that I could create a new problem just by changing the ID, and then started doing that. But I can't tell you which of the problems I created which way. I am pretty sure that when I created new problems by editing an id it was based on one of the problems I had written, not one that already existed in the book.

**Comment by bnmnetp** [comment]
Ok, multiple issues here, but I think I have a handle on them.  And I'm glad that we are finding these before my classes start!!

1.  I thought the templates that included all of the options would be a nice way to get people started writing their own questions.  That appears to not be true.  Would a simple example to start from be better?

2.  I think the fact that the problems were not showing up in the detail was caused by the lack of defaults.  That has been corrected.  Or at least I can't reproduce that at any rate.

3.  I have also uncovered a problem that @haugrud reported  RunestoneServer#1013   -- So you autograde --> manual grade --> then look at the grade book.  The scores for the manual grades are missing until you go back and do another auto grade.  -- I am not updating the assignment total after the manual grade. -- I'll need to fix that.  Probably can do that yet today.

Thanks for your patience and persistence in helping track this stuff down!

**Comment by cigas** [comment]
At least for now, I think a simple example would be best. 

I actually liked the thought of the complete template until I started trying to use it. Just in the first 3 lines are 3 different elements
.. activecode:: uniqueid                               - Here uniqueid is something the author has to supply
   :nocanvas: do not create a canvas           - I think "do not create..." is just a comment
   :autograde: normally set this to unittest  - I don't know where to put the word 'unittest'

I hate to say it, but you probably need a more formal way of describing all the possibilities when you do document this.

It was also problematic for me that when I went to an already designed exercise and copied the code,
.. actex:: ex_2_8
and then changed the words. None of my own text would display. I realize now that the problem was not changing the ex_2_8 to something different, but at the time, I thought it was a problem with the term 'actex' vs. 'activecode'. It seems like a lot of work, but it would be nice if there were only one term to use. 

Also, if you can add a pop up when someone besides the original author tries to reuse a uniqueid you'll save other people a lot of head scratching.

**Comment by bnmnetp** [comment]
Yes, I've just done that.  If you are not the author of the question you cannot save it.  If you try to reuse a unique id that is already in use by someone else you cannot save.

We're getting there...
